sql: don't skip lower bytes of unicode points

madelynnblue opened the pull request:
Iterate over all bytes, not just unicode code points.

This is hopefully the last bug fix for this function. The previous
iteration de-duplicated some things, but still didn't catch the unprintable
characters. Now we have an explicit test to prevent any weird bytes.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7350)

<!-- Reviewable:end -->

mjibson commented:
One more of these -_-

knz commented:
Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[sql/parser/encode.go, line 96 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7350#-KKlp5JW9TYJeinGfbsv:-KKlp5JXcJX9Unah574x:-1767435338) ([raw file](https://github.com/cockroachdb/cockroach/blob/2213e2bb5f1fe59de3110fd839696230498bb8ab/sql/parser/encode.go#L96)):*
> ```Go
> 	start := 0
> 	buf.WriteString("b'")
> 	for i, n := 0, len(in); i < n; i++ {
> ```

I do not understand how this is different from `i := range in`. Since neither in nor n is modified in the loop, it should be the same right?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7350)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[sql/parser/encode.go, line 96 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7350#-KKlp5JW9TYJeinGfbsv:-KKlzVRmA0v3IKSdznXM:1596290681) ([raw file](https://github.com/cockroachdb/cockroach/blob/2213e2bb5f1fe59de3110fd839696230498bb8ab/sql/parser/encode.go#L96)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> I do not understand how this is different from `i := range in`. Since neither in nor n is modified in the loop, it should be the same right?
</details>
`range` over a string returns the unicode code points, not each byte see: https://blog.golang.org/strings#TOC_6.).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7350)*
<!-- Sent from Reviewable.io -->

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[sql/parser/encode.go, line 96 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7350#-KKlp5JW9TYJeinGfbsv:-KKm8oS_JCFo0ITtuzWr:735629216) ([raw file](https://github.com/cockroachdb/cockroach/blob/2213e2bb5f1fe59de3110fd839696230498bb8ab/sql/parser/encode.go#L96)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> `range` over a string returns the unicode code points, not each byte see: https://blog.golang.org/strings#TOC_6.).
</details>
Ugh. And len() returns the number of bytes, not characters. This smells :cry:

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7350#-:-KKm9OhUC4oJJUAbzZKj:1417088721)*
<!-- Sent from Reviewable.io -->

paperstreet commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[sql/parser/encode.go, line 96 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7350#-KKlp5JW9TYJeinGfbsv:-KKnSXdqRdF9598cYtSx:2056962198) ([raw file](https://github.com/cockroachdb/cockroach/blob/2213e2bb5f1fe59de3110fd839696230498bb8ab/sql/parser/encode.go#L96)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Ugh. And len() returns the number of bytes, not characters. This smells :cry:
</details>
Worth a comment?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7350)*
<!-- Sent from Reviewable.io -->

madelynnblue committed:
sql: don't skip lower bytes of unicode points

Iterate over all bytes, not just unicode code points.

This is hopefully the last bug fix for this function. The previous
iteration de-duplicated some things, but still didn't catch the unprintable
characters. Now we have an explicit test to prevent any weird bytes.

mjibson commented:
Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[sql/parser/encode.go, line 96 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7350#-KKlp5JW9TYJeinGfbsv:-KKnvD2sfpzNUgFSL0LW:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/2213e2bb5f1fe59de3110fd839696230498bb8ab/sql/parser/encode.go#L96)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> Worth a comment?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7350)*
<!-- Sent from Reviewable.io -->
